enhancements/cluster-logging/logs-based-alerts.md

```diff
@@ -0,0 +1,300 @@
+- cluster admins can configure how the platform logs can notify on alertworthy platform and/or audit logs.
+- application owners can configure how the workload logs can notify on alertworthy logs.
+
+### Workflow Description
+
+#### Logs-based Alerts for Application Owners
+
+1. The application owner creates an AlertingRule custom resource on a workload project, e.g.
+
+```yaml
+apiVersion: loki.grafana.com/v1beta1
+kind: AlertingRule
+metadata:
+  name: my-workload-alerts
+  namespace: my-workload-ns
+  labels:
+    openshift.io/cluster-monitoring: "true"
+spec:
+  tenantID: application
+  groups:
+    - name: MyApplication
+      rules:
+        - alert: MyApplicationHighPercentageError
+          expr: |
+            sum(rate({kubernetes_namespace_name="my-workload-ns", kubernetes_pod_name=~"my-workload.*"} |= "error" [1m])) by (job)
+              /
+            sum(rate({kubernetes_namespace_name="my-workload-ns", kuberentes_pod_name=~"my-workload.*"}[1m])) by (job)
+              > 100
+          for: 5m
+          labels:
+            severity: warning
+          annotations:
+            summary: My Application High Errors.
+            description: My application throws a high amount of error logs.
+```
+
+2. The application owner can only define AlertingRules custom resources in project with granted access to them and only for the tenant `application`.
+3. The application owner can inspect firing alerts from the above rule in Developer > Observe > Alerts.
+
+#### Logs-based Alerts for Cluster Admins
+
+1. The cluster admin creates an AlertingRule custom resource on a platform project, e.g.
+
+```yaml
+apiVersion: loki.grafana.com/v1beta1
+kind: AlertingRule
+metadata:
+  name: my-operator-alerts
+  namespace: openshift-operators-redhat
+  labels:
+    openshift.io/cluster-monitoring: "true"
+spec:
+  tenantID: infrastructure
+  groups:
+    - name: MyOperator
+      rules:
+        - alert: MyOperatorHighPercentageError
+          expr: |
+            sum(rate({kubernetes_namespace_name="openshift-operators-redhat", kubernetes_pod_name=~"my-operator.*"} |= "error" [1m])) by (job)
+              /
+            sum(rate({kubernetes_namespace_name="openshift-operators-redhat", kuberentes_pod_name=~"my-operator.*"}[1m])) by (job)
+              > 100
+          for: 5m
+          labels:
+            severity: warning
+          annotations:
+            summary: My Operator High Errors.
+            description: My operator throws a high amount of error logs.
+```
+
+2. The cluster admin can define AlertingRule custom resources all projects for tenants `application`, `infrastructure` and `audit`
+3. The cluster admin can inspect firing alerts from the above rule in Administrator > Observe > Alerting > Alerts.
+
+### API Extensions
+
+As mentioned above this document proposes to leverage two existing CRDs namely `AlertingRule` and `RulerConfig` provided by the Loki Operator.
+
+### Implementation Details/Notes/Constraints
```

Comment:
Until now we've treated the Alertmanager deployment as an implementation detail and didn't make any promise on the stability (e.g. we could change the authentication method or require different permissions if necessary). If the Loki Ruler becomes another source of alerts, we would need to be more careful (not saying it's bad but this needs to be taken into account on the monitoring side). Something that comes to my mind for instance: in 4.11 we're adding the possibility to deploy a standalone Alertmanager for user-defined alerts so the Alertmanager destination for alerts could be the Alertmanager service in openshift-monitoring namespace or the Alertmanager service in the openshift-user-workload-monitoring namespace.

Reply:
Yeah exactly this is main reason why I want to raise this as an enhancement proposal first. Placing the Loki Ruler as extra notifier will basically raise the levels on an OCP in-cluster component like Alertmanager in terms of availability. To push this a little bit further, we will have the same issue with Prometheus when Loki's recording rules will use the remote-write-client to push metrics to cluster-monitoring's/user-workload-monitoring's prometheus. But let's frame this for AlertManager first.

My questions to you folks are:
- What does it take to make raise availability of alertmanager as a component that can be used from components like Loki ruler that are not part of OCP but belong to a layered product like OpenShift Logging?
- How can we detect if users have one Alertmanager (cluster-monitoring) or more Alertmanagers (cluster-monitoring + user-workload-monitoring)?